src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.HandleNull.cs

```diff
@@ -0,0 +1,484 @@
+        {
+            // Baseline
+            int? val = JsonSerializer.Deserialize<int?>("null");
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val));
+
+            // Per null handling default value for value types (true), converter handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new NullableInt32NullConverter_SpecialCaseNull());
+
+            val = JsonSerializer.Deserialize<int?>("null", options);
+            Assert.Equal(-1, val);
+
+            val = null;
+            Assert.Equal("-1", JsonSerializer.Serialize(val, options));
+        }
+
+        private class NullableInt32NullConverter_SpecialCaseNull : JsonConverter<int?>
+        {
+            public override int? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.TokenType == JsonTokenType.Null)
+                {
+                    return -1;
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(Utf8JsonWriter writer, int? value, JsonSerializerOptions options)
+            {
+                if (!value.HasValue)
+                {
+                    writer.WriteNumberValue(-1);
+                    return;
+                }
+
+                throw new NotSupportedException();
+            }
+        }
+
+        [Fact]
+        public static void NullableValueTypeConverter_OptOut()
+        {
+            // Baseline
+            int? val = JsonSerializer.Deserialize<int?>("null");
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val));
+
+            // Per null handling opt-out, serializer handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new NullableInt32NullConverter_NullOptOut());
+
+            val = JsonSerializer.Deserialize<int?>("null", options);
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val, options));
+        }
+
+        private class NullableInt32NullConverter_NullOptOut : NullableInt32NullConverter_SpecialCaseNull
+        {
+            public override bool HandleNull => false;
+        }
+
+        [Fact]
+        public static void ReferenceTypeConverter_NoOverride()
+        {
+            // Baseline
+            Uri val = JsonSerializer.Deserialize<Uri>("null");
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val));
+
+            // Per null handling default value for reference types (false), serializer handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new Int32NullConverter_SpecialCaseNull());
+
+            val = JsonSerializer.Deserialize<Uri>("null", options);
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val, options));
```

Comment:
"https://default" instead?

Or are you checking to see if we call the correct converter? Is so, please comment.

Reply:
Here, the reference type converter doesn't override `HandleNull`, so the property is false. The serializer will set the default value when deserializing and write `null` when serializing. I've added a comment.